upgrader.go

```diff
@@ -70,120 +74,131 @@ func WithLogger(l log15.Logger) Option {
-				return errClosed
+				u.l.Info("upgrade socket closed, no longer listening for upgrades")
+				return
 			}
-			return errors.Wrap(err, "error accepting upgrade socket request")
-		}
-		defer conn.Close()
-
-		// We got a request, only handle one request at a time via semaphore..
-		// Acquire semaphore, but don't block. This allows informing
-		// the user that they are doing too many upgrade requests.
-		select {
-		default:
-			return errors.New("upgrade in progress")
-		case u.upgradeSem <- struct{}{}:
+			u.l.Error("error awaiting upgrade", "err", err)
+			continue
 		}
+		go u.handleUpgradeRequest(ctx, conn)
+	}
+}
 
-		defer func() {
-			<-u.upgradeSem
-		}()
+func (u *Upgrader) transitionTo(state upgraderState) error {
+	u.stateLock.Lock()
+	defer u.stateLock.Unlock()
+	return u.state.transitionTo(state)
+}
 
-		// Make sure we're still ok to perform an upgrade.
-		select {
-		case <-u.exitC:
-			return errors.New("already upgraded")
-		default:
-		}
+func (u *Upgrader) mustTransitionTo(state upgraderState) {
+	u.stateLock.Lock()
+	defer u.stateLock.Unlock()
+	if err := u.state.transitionTo(state); err != nil {
+		panic(fmt.Sprintf("BUG: error transitioning to %q: %v", state, err))
+	}
+}
 
-		select {
-		case <-u.readyC:
-		default:
-			return errors.New("this process cannot service an upgrade request until it is ready; not yet marked ready")
-		}
+func (u *Upgrader) handleUpgradeRequest(ctx context.Context, conn *net.UnixConn) {
+	defer conn.Close()
 
-		u.l.Info("passing along the torch")
-		// time to pass our FDs along
-		nextParent, errC := passFdsToSibling(u.l, conn, u.Fds.copy())
+	if err := u.transitionTo(upgraderStateTransferringOwnership); err != nil {
+		u.l.Info("cannot handle upgrade request", "reason", err)
+		return
+	}
 
-		readyTimeout := time.After(u.upgradeTimeout)
-		select {
-		case err := <-errC:
-			return fmt.Errorf("next parent gave us an error: %v", err)
-		case <-u.stopC:
-			return errors.New("terminating")
-		case <-readyTimeout:
-			return errors.Errorf("new parent %s timed out", nextParent)
-		case <-nextParent.readyC:
-			u.l.Info("next parent is ready, marking ourselves as up for exit")
-			close(u.exitC)
-			return nil
+	u.l.Info("handling an upgrade request from peer")
+	// time to pass our FDs along
+	nextOwner, errC := passFdsToSibling(u.l, conn, u.Fds.copy())
+
+	readyTimeout := time.NewTimer(u.upgradeTimeout)
+	defer readyTimeout.Stop()
+	select {
+	case err := <-errC:
+		u.l.Error("failed to pass file descriptors to next owner", "reason", "error", "err", err)
+		// remain owner
+		if err := u.transitionTo(upgraderStateOwner); err != nil {
+			// could happen if 'Stop' was called after 'handleUpgradeRequest'
+			// started, and then the request failed.
+			// This leaves us in the state of being the sole owner of Fds, but not
+			// being able to pass on ownership because that's what 'Stop' indicates
+			// is desired.
+			// At this point, we can't really do anything but complain.
```

Comment:
presumably this could also happen if the process initiating the ownership transfer failed in some way as well. (doesn't change the code, but might want to be noted in the comment)

Reply:
The error channel cold get an error in that case, yes, but transitioning this process from `transferringOwnership` back to `Owner` shouldn't be able to fail other than in the case of `Stop`. Anything that messed up in the sibling process can't impact the internal state of our upgrader.

This comment was just meant to be about state transitions.